src/rtprecv.c

```diff
@@ -0,0 +1,503 @@
+		}
+	}
+	else {
+		debug("stream: receive: setting SSRC: %x\n", ssrc);
+		rx->ssrc = ssrc;
+		rx->ssrc_set = true;
+	}
+	mtx_unlock(rx->mtx);
+}
+
+
+uint64_t rtprecv_ts_last(struct rtp_receiver *rx)
+{
+	uint64_t ts_last;
+	mtx_lock(rx->mtx);
+	ts_last = rx->ts_last;
+	mtx_unlock(rx->mtx);
+
+	return ts_last;
+}
+
+
+void rtprecv_set_ts_last(struct rtp_receiver *rx, uint64_t ts_last)
+{
+	mtx_lock(rx->mtx);
+	rx->ts_last = ts_last;
+	mtx_unlock(rx->mtx);
+}
+
+
+void rtprecv_flush(struct rtp_receiver *rx)
+{
+	if (!rx)
+		return;
+
+	jbuf_flush(rx->jbuf);
+}
+
+
+void rtprecv_set_enable(struct rtp_receiver *rx, bool enable)
+{
+	if (!rx)
+		return;
+
+	mtx_lock(rx->mtx);
+	rx->enabled = enable;
+	mtx_unlock(rx->mtx);
+}
+
+
+int rtprecv_get_ssrc(struct rtp_receiver *rx, uint32_t *ssrc)
+{
+	int err;
+
+	if (!rx || !ssrc)
+		return EINVAL;
+
+	mtx_lock(rx->mtx);
+	if (rx->ssrc_set) {
+		*ssrc = rx->ssrc;
+		err = 0;
+	}
+	else
+		err = ENOENT;
+	mtx_unlock(rx->mtx);
+
+	return err;
+}
+
+
+void rtprecv_enable_mux(struct rtp_receiver *rx, bool enable)
+{
+	mtx_lock(rx->mtx);
+	rtcp_enable_mux(rx->rtp, enable);
+	mtx_unlock(rx->mtx);
+}
+
+
+/**
+ * The debug function prints into the given mbuf in order to avoid long
+ * blocking print to stdout.
+ *
+ * @param rx The rtp_receiver
+ * @param mb Memory buffer
+ *
+ * @return 0 if success, otherwise errorcode
+ */
+int rtprecv_debug(struct re_printf *pf, const struct rtp_receiver *rx)
+{
+	int err;
+	bool enabled;
+
+	mtx_lock(rx->mtx);
+	enabled = rx->enabled;
+	mtx_unlock(rx->mtx);
+
+	err  = re_hprintf(pf, " rx.enabled: %s\n", enabled ? "yes" : "no");
+	err |= jbuf_debug(pf, rx->jbuf);
+
+	return err;
+}
+
+
+static void destructor(void *arg)
+{
+	struct rtp_receiver *rx = arg;
+
+	mem_deref(rx->metric);
+	mem_deref(rx->name);
+	mem_deref(rx->mtx);
+	mem_deref(rx->jbuf);
+}
+
+
+int rtprecv_alloc(struct rtp_receiver **rxp,
+		   struct stream *strm,
+		   const char *name,
+		   const struct config_avt *cfg,
+		   stream_rtp_h *rtph,
+		   stream_pt_h *pth, void *arg)
+{
+	struct rtp_receiver *rx;
+	int err;
+
+	if (!rxp || !str_isset(name))
+		return EINVAL;
+
+	rx = mem_zalloc(sizeof(*rx), destructor);
+	if (!rx)
+		return ENOMEM;
+
+	MAGIC_INIT(rx);
+	rx->strm   = strm;
+	rx->rtph   = rtph;
+	rx->pth    = pth;
+	rx->arg    = arg;
+	rx->pseq   = -1;
+	rx->pt     = -1;
+	err  = str_dup(&rx->name, name);
+	err |= mutex_alloc(&rx->mtx);
+
+	/* Audio Jitter buffer */
+	if (stream_type(strm) == MEDIA_AUDIO &&
+	    cfg->audio.jbtype != JBUF_OFF && cfg->audio.jbuf_del.max) {
+
+		err = jbuf_alloc(&rx->jbuf, cfg->audio.jbuf_del.min,
+				 cfg->audio.jbuf_del.max);
+		err |= jbuf_set_type(rx->jbuf, cfg->audio.jbtype);
+	}
+
+	/* Video Jitter buffer */
+	if (stream_type(strm) == MEDIA_VIDEO &&
+	    cfg->video.jbtype != JBUF_OFF && cfg->video.jbuf_del.max) {
+
+		err = jbuf_alloc(&rx->jbuf, cfg->video.jbuf_del.min,
+				 cfg->video.jbuf_del.max);
+		err |= jbuf_set_type(rx->jbuf, cfg->video.jbtype);
```

Comment:
You should add a explicit `if (err) goto out;` right after `jbuf_alloc`, otherwise you get `ENOMEM(12) | EINVAL(22)` = misleading EROFS(30) errors. jbuf_set_type return code can be ignored.